File: LeetCode/Medium/347. Top K Frequent Elements.py
# Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

 

# Example 1:

# Input: nums = [1,1,1,2,2,3], k = 2
# Output: [1,2]
# Example 2:

# Input: nums = [1], k = 1
# Output: [1]
 

# Constraints:

# 1 <= nums.length <= 105
# -104 <= nums[i] <= 104
# k is in the range [1, the number of unique elements in the array].
# It is guaranteed that the answer is unique.
 

# Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.

# My Python Solution O(n) 
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """

        count_nums: dict[int, int] = defaultdict(int)
        max_k: int = 0

        for num in nums:
            count_nums[num] += 1
            max_k = max(max_k, count_nums[num])

        by_count = [[] for i in range(max_k + 1)]

        for num, count in count_nums.items():
            print(num,count)
            by_count[count].append(num)


        freq_k = []
        iterations = k

        for i in range(max_k, 0, -1):
            for num in by_count[i]:
                if iterations:
                    freq_k.append(num)
                    iterations -= 1


        return freq_k


# Code Wars Alternative
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # using heapq
        count  = Counter(nums)
        res = heapq.nlargest(k, count, key = count.get)
        return res

File: LeetCode/Medium/347. Top K Frequent Elements.cpp
// Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

 

// Example 1:

// Input: nums = [1,1,1,2,2,3], k = 2
// Output: [1,2]
// Example 2:

// Input: nums = [1], k = 1
// Output: [1]
 

// Constraints:

// 1 <= nums.length <= 105
// -104 <= nums[i] <= 104
// k is in the range [1, the number of unique elements in the array].
// It is guaranteed that the answer is unique.
 

// Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.

// My C++ Solution in O(n)

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        std::unordered_map<int,int> count;


        for(int num : nums) {
            count[num]++;
        }

        int cap = nums.size();
        vector<vector<int>> bucket(cap + 1);
        for(pair<int, int> pair : count) {
            bucket[pair.second].push_back(pair.first);
        }

        vector<int> res;
        int iteration = 0;

        for(int i = cap; i >= 0 ; i--) {
            for(int num : bucket[i]){
                if(res.size() == k) break;
                res.push_back(num);

            }


        }
        return res;
    }
};

// LeetCode Community Alternative
class Solution {
public:
     std::vector<int> topKFrequent(std::vector<int>& nums, int k) {

          std::unordered_map<int, int> hashmap{};
          for (const auto& i : nums)
               ++hashmap[i];

          std::vector<std::pair<int, int>> vec(begin(hashmap), end(hashmap));

          std::sort(begin(vec), end(vec), [](const auto& x, const auto& y)
               {return y.second < x.second; });

          std::vector<int> top{};
          top.reserve(k);
          for (int i{ 0 }; i < k; ++i)
               top.push_back(vec[i].first);

          return top;

     }
};
